File: backend/api/app/schemas/companion_doctor.py
from typing import Literal, Optional

from pydantic import BaseModel, Field


class CompanionStartRequest(BaseModel):
    language_preference: str = Field(default="auto")
    mode: Literal["voice", "text"] = Field(default="voice")


class CompanionStartResponse(BaseModel):
    session_id: str
    assistant_text: str
    detected_language: str
    safety_flags: list[str] = Field(default_factory=list)
    emergency_flag: bool = False
    should_end_session: bool = False
    remaining_free_turns: int
    state: str
    privacy_note: str
    non_emergency_notice: str
    access_locked: bool = False
    upgrade_url: Optional[str] = None


class CompanionMessageRequest(BaseModel):
    session_id: str
    message: str = Field(min_length=1, max_length=1500)
    language_preference: str = Field(default="auto")
    channel: Literal["voice", "text"] = Field(default="text")


class CompanionMessageResponse(BaseModel):
    session_id: str
    detected_language: str
    session_language: Optional[str] = None      # locked language for the whole session
    refined_user_text: str
    assistant_text: str
    tts_text: str
    safety_flags: list[str] = Field(default_factory=list)
    emergency_flag: bool = False
    emotion: Optional[str] = None               # listening | calm | concerned | reassuring | urgent
    asked_slot: Optional[str] = None            # which intake slot was targeted this turn
    safety_level: Optional[str] = None          # normal | caution | urgent
    should_end_session: bool = False
    remaining_free_turns: int
    state: str
    access_locked: bool = False
    upgrade_url: Optional[str] = None


class CompanionEndRequest(BaseModel):
    session_id: str


class CompanionEndResponse(BaseModel):
    ok: bool
    session_id: str
    message: str


class CompanionFeedbackRequest(BaseModel):
    session_id: Optional[str] = None
    rating: int = Field(ge=1, le=5)
    sentiment: Optional[Literal["up", "down"]] = None
    text: Optional[str] = Field(default=None, max_length=1500)
    consent_to_share: bool = False


class CompanionFeedbackResponse(BaseModel):
    ok: bool
    stored: bool
    message: str


class CompanionSTTResponse(BaseModel):
    session_id: Optional[str] = None
    raw_transcript: str
    refined_user_text: str
    detected_language: str
    safety_flags: list[str] = Field(default_factory=list)
    emergency_flag: bool = False


class CompanionTTSRequest(BaseModel):
    session_id: Optional[str] = None
    text: str = Field(min_length=1, max_length=3000)
    language_preference: str = Field(default="auto")
    voice: Optional[str] = None
    provider: Optional[str] = None

File: frontend/types/companion-doctor.ts
export type CompanionMode = "voice" | "text";
export type CompanionStatus = "idle" | "listening" | "thinking" | "speaking";
export type CompanionLanguage = "auto" | "en" | "hi" | "mixed";
export type CompanionThumb = "up" | "down";

export type CompanionConversationItem = {
  id: string;
  role: "assistant" | "user";
  text: string;
  language: CompanionLanguage | string;
};

export type CompanionStartResponse = {
  session_id: string;
  assistant_text: string;
  detected_language: CompanionLanguage;
  safety_flags: string[];
  emergency_flag: boolean;
  should_end_session: boolean;
  remaining_free_turns: number;
  state: string;
  privacy_note: string;
  non_emergency_notice: string;
  access_locked: boolean;
  upgrade_url?: string | null;
};

export type CompanionMessageResponse = {
  session_id: string;
  detected_language: CompanionLanguage;
  refined_user_text: string;
  assistant_text: string;
  tts_text: string;
  safety_flags: string[];
  emergency_flag: boolean;
  should_end_session: boolean;
  remaining_free_turns: number;
  state: string;
  access_locked: boolean;
  upgrade_url?: string | null;
};

export type CompanionSTTResponse = {
  session_id?: string | null;
  raw_transcript: string;
  refined_user_text: string;
  detected_language: CompanionLanguage;
  safety_flags: string[];
  emergency_flag: boolean;
};

export type CompanionFeedbackResponse = {
  ok: boolean;
  stored: boolean;
  message: string;
};

export type CompanionFeedbackDraft = {
  rating: number;
  sentiment: CompanionThumb | null;
  text: string;
  consentToShare: boolean;
};